### PR TITLE
New Data Source:  aws_s3_bucket_policy

### LINKED ISSUE #17649: aws_s3_bucket_policy data source
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Re-opening a very old request (https://github.com/hashicorp/terraform-provider-aws/issues/6977) for the S3 Bucket Policy data-source as I feel it warrants an actual discussion request.

We have a few situations where we'd like to be able to reference an existing S3 Bucket Policy to append an new Statement into it then apply that policy 'back' to the S3 bucket (using aws_iam_policy_document and source_json to perform the merge).

Currently we're just using a separate S3 Object to 'mirror' this S3 Bucket Policy and reading/writing to that object before/after the Bucket Policy application, but this is hacky and involves extra steps and increases the chances of failure!

### New or Affected Resource(s)

aws_s3_bucket_policy

### Potential Terraform Configuration

```hcl
data "aws_s3_bucket_policy" "policy" {
  arn = "arn:aws:s3:::my-bucket-name"
}

data "aws_s3_bucket_policy" "policy" {
  name = "my-bucket-name"
}
```

### PR DESCRIPTION
**Pull request by yuonoda**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #17649.
Relates #6334.

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceS3BucketPolicy_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceS3BucketPolicy_basic -timeout 120m
=== RUN   TestAccDataSourceS3BucketPolicy_basic
=== PAUSE TestAccDataSourceS3BucketPolicy_basic
=== CONT  TestAccDataSourceS3BucketPolicy_basic
--- PASS: TestAccDataSourceS3BucketPolicy_basic (94.96s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       99.214s
```

### CODE REVIEW

File: website/docs/d/s3_bucket_policy.html.markdown [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+The following example retrieves IAM policy of a specified S3 bucket.
+
+```hcl
+data "aws_s3_bucket_policy" "policy" {
+  bucket = "example-bucket-name"
+}
+
+output "foo" {
+  value = data.aws_s3_bucket_policy.policy
```
Comment (svanzoest) [inline website/docs/d/s3_bucket_policy.html.markdown:22]: should this be `data.aws_s3_bucket_policy.policy.policy`?

**Comment by johnsonaj** [comment]
#### Commercial
```console
% make testacc TESTS=TestAccDataSourceBucketPolicy_ PKG=s3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/s3/... -v -count 1 -parallel 20 -run='TestAccDataSourceBucketPolicy_'  -timeout 180m
=== RUN   TestAccDataSourceBucketPolicy_basic
=== PAUSE TestAccDataSourceBucketPolicy_basic
=== CONT  TestAccDataSourceBucketPolicy_basic
--- PASS: TestAccDataSourceBucketPolicy_basic (16.39s)
PASS
```

#### Gov cloud
```console
% make testacc TESTS=TestAccDataSourceBucketPolicy_ PKG=s3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/s3/... -v -count 1 -parallel 20 -run='TestAccDataSourceBucketPolicy_'  -timeout 180m
=== RUN   TestAccDataSourceBucketPolicy_basic
=== PAUSE TestAccDataSourceBucketPolicy_basic
=== CONT  TestAccDataSourceBucketPolicy_basic
--- PASS: TestAccDataSourceBucketPolicy_basic (14.64s)
PASS
```

**Review by johnsonaj — APPROVED** [review]

**Comment by johnsonaj** [comment]
@yuonoda Thanks for the contribution! 🎉 👏🏾

**Comment by yermulnik** [comment]
@yuonoda Could there be an option to not fail but return an empty object (along with a warning message) for when S3 bucket has no bucket policy attached?
My use case is to amend policy of manually created S3 bucket hence I use `source_policy_documents` parameter within `aws_iam_policy_document` resource and the input for this parameter is retrieved by means of `aws_s3_bucket_policy` data source which obviously fails with `Error: failed getting S3 bucket policy (my-cool-bucket-name): couldn't find resource` when bucket policy doesn't exist at all.
Thanks

**Pull request merged**